core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -942,6 +969,9 @@ public void run() {
           } else {
             RetryPlan retryPlan = makeRetryDecision(status, trailers);
             if (retryPlan.shouldRetry) {
+              if (inFlightSubStreams.incrementAndGet() < 0) {
```

Comment:
This needs to happen in RetryBackoffRunnable.run(), before callExecutor.execute(), as otherwise we will delay cleanup even if this retry never happens. We should probably move the createSubstream() out of the callExecutor.execute() to make the relationship more clear, and because it is safe to call on the current thread.